controllers/tls/manager.go

```diff
@@ -0,0 +1,342 @@
+		return fmt.Errorf("missing %s field in %s secret", corev1.ServiceAccountRootCAKey, r.Configuration.TLSSecretName())
+	}
+
+	operatorPods, err := r.getOperatorPods(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.Log.Info("Updating caBundle in webhooks and crd")
+
+	group := new(errgroup.Group)
+	group.Go(func() error {
+		return r.updateMutatingWebhookConfiguration(ctx, caBundle)
+	})
+	group.Go(func() error {
+		return r.updateValidatingWebhookConfiguration(ctx, caBundle)
+	})
+	group.Go(func() error {
+		return r.updateCustomResourceDefinition(ctx, caBundle)
+	})
+
+	r.Log.Info("Updating capsule operator pods")
+
+	for _, pod := range operatorPods.Items {
+		p := pod
+
+		group.Go(func() error {
+			return r.updateOperatorPod(ctx, p)
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	r.Log = r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+
+	certSecret := &corev1.Secret{}
+
+	if err := r.Client.Get(ctx, request.NamespacedName, certSecret); err != nil {
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if err := r.ReconcileCertificates(ctx, certSecret); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if r.Configuration.GenerateCertificates() {
+		certificate, err := cert.GetCertificateFromBytes(certSecret.Data[corev1.TLSCertKey])
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		now := time.Now()
+		requeueTime := certificate.NotAfter.Add(-(certificateExpirationThreshold - 1*time.Second))
+		rq := requeueTime.Sub(now)
+
+		r.Log.Info("Reconciliation completed, processing back in " + rq.String())
+
+		return reconcile.Result{Requeue: true, RequeueAfter: rq}, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r Reconciler) shouldUpdateCertificate(secret *corev1.Secret) bool {
+	if !r.Configuration.GenerateCertificates() {
+		r.Log.Info("Skipping TLS certificate generation as it is disabled in CapsuleConfiguration")
+
+		return false
+	}
+
+	if _, ok := secret.Data[corev1.ServiceAccountRootCAKey]; !ok {
+		return true
+	}
+
+	certificate, key, err := cert.GetCertificateWithPrivateKeyFromBytes(secret.Data[corev1.TLSCertKey], secret.Data[corev1.TLSPrivateKeyKey])
+	if err != nil {
+		return true
+	}
+
+	if err := cert.ValidateCertificate(certificate, key, certificateExpirationThreshold); err != nil {
+		r.Log.Error(err, "failed to validate certificate, generating new one")
+
+		return true
+	}
+
+	r.Log.Info("Skipping TLS certificate generation as it is still valid")
+
+	return false
+}
+
+// By default helm doesn't allow to use templates in CRD (https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you).
+// In order to overcome this, we are setting conversion strategy in helm chart to None, and then update it with CA and namespace information.
```

Comment:
Now we got CRD in the `templates` folder, this comment is outdated.

Reply:
nope :) as we discussed in slack, moving `crds` to `template` folder will cause this error
```
Error: UPGRADE FAILED: rendered manifests contain a resource that already exists. Unable to continue with update: CustomResourceDefinition "tenants.capsule.clastix.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "capsule"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "capsule-system"
helm.go:84: [debug] CustomResourceDefinition "tenants.capsule.clastix.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "capsule"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "capsule-system"
rendered manifests contain a resource that already exists. Unable to continue with update
```
when updating, so this will be backward-incompatible change. That's why we still keep them in `crds` and update `caBundle` for webhook manually